[CDAP-9532] Adds Encode and Decode point and click directives to dataprep

ajainarayanan opened the pull request:
- Adds EncodeDecode component to handle both encoding and decoding data in dataprep 
- Disables certain decode options that require the column to be of type `byte[]`.

JIRA: https://issues.cask.co/browse/CDAP-9532

ajainarayanan committed:
[CDAP-9532] Adds Encode and Decode point and click directives to dataprep

 - Adds a genric EncodeDecode and made Decode a special case for reusability
 - Fixes setPopoverOffset helper function to fix bottom of second level popover in dataprep
 - Adds an additional disabled state for Decode options if the column type is not byte[]
 - Fixes second level popover to not hover over the footer + fixes second level popover top when position bottom relative to the main menu item
 - Moves char encoding options to a separate menu, and shows tooltip when column type is not byte array

tonybach commented:
<img width="1265" alt="screen shot 2017-06-13 at 4 01 09 pm" src="https://user-images.githubusercontent.com/6516002/27108382-b9c3dacc-5051-11e7-9f14-9404bb2b457b.png">

Not exactly related to this PR, but this is what a maximum-height window looks like on my 13-inch Macbook without going full screen. I can't even see Encode or Decode at all. Not sure how to handle this though, since it's inevitable that this will happen as we add more directives :( 

**EDIT**: Oh just noticed there's https://issues.cask.co/browse/CDAP-11886.

tonybach commented on cdap-ui/app/cdap/components/DataPrep/Directives/EncodeDecode/EncodeDecode.scss:
```diff
@@ -0,0 +1,35 @@
+        pointer-events: auto;
+      }
+      &:hover {
+        background-color: #dddddd;
```
Extract this color?

tonybach commented:
One minor comment, otherwise code LGTM 👍

ajainarayanan commented:
cc: @elfenheart I have modified calculating second level popover offset to align with the main menu item. Let me know if I missed a scenario because I have removed the footer height while calculating the body bottom.

elfenheart commented:
@ajainarayanan do we want it to overflow over the footer?

elfenheart reviewed (approved):
Just a question, otherwise 👍  LGTM

ajainarayanan commented:
@elfenheart Have modified it so that the second level popover doesn't hover on top of footer. @tonybach Have fixed the weird second level popover issue you are seeing. Reg the popover going beyond the viewport have another PR that addresses that: https://github.com/caskdata/cdap/pull/9109

bdmogal commented:
@tonybach let's add a horizontal rule between the various decoding params and character sets?
With a heading "Set Charset"

tonybach commented:
Discussed with @bdmogal, will move `UTF-8`, `UTF-16`, `US-ASCII` etc. from the `Decode` menu and move it to a new one called `Set Character Encoding`.

elfenheart commented on cdap-ui/app/cdap/components/DataPrep/Directives/EncodeDecode/index.js:
```diff
@@ -117,20 +113,31 @@ export default class EncodeDecode extends Component {
+        </div>
+        {
+          this.props.isDisabled && this.props.disabledTooltip ? (
+            <UncontrolledTooltip target={id} delay={{show: 250, hide: 0}}>
```
new line for each of the attributes

elfenheart reviewed (approved):
👍  LGTM

ajainarayanan commented:
Have a LGTM from @bdmogal . Merging.
